solr/core/src/java/org/apache/solr/util/tracing/SimplePropagator.java

```diff
@@ -0,0 +1,304 @@
+
+  private static final AtomicLong traceCounter = new AtomicLong(0);
+
+  public static void load() {
+    GlobalTracer.registerIfAbsent(
+        () -> {
+          log.info("Always-on trace id generation enabled.");
+          return new SimplePropagatorTracer();
+        });
+  }
+
+  /**
+   * Tracer that only aims to do simple header propagation, tailored to how Solr works.
+   *
+   * <p>Heavily inspired from JaegerTracer, NoopTracer
+   */
+  static class SimplePropagatorTracer implements Tracer {
+
+    private final ScopeManager scopeManager = new ThreadLocalScopeManager();
+
+    @Override
+    public ScopeManager scopeManager() {
+      return scopeManager;
+    }
+
+    @Override
+    public Span activeSpan() {
+      return scopeManager.activeSpan();
+    }
+
+    @Override
+    public Scope activateSpan(Span span) {
+      return scopeManager.activate(span);
+    }
+
+    @Override
+    public SpanBuilder buildSpan(String operationName) {
+      return new SimplePropagatorSpanBuilder(scopeManager);
+    }
+
+    @Override
+    public void close() {}
+
+    @Override
+    public <C> void inject(SpanContext spanContext, Format<C> format, C carrier) {
+      if (!format.equals(Format.Builtin.HTTP_HEADERS)) {
+        throw new IllegalArgumentException("unsupported format " + format);
+      }
+      String traceId = spanContext.toTraceId();
+      if (traceId != null && !traceId.isEmpty()) {
+        TextMap tm = (TextMap) carrier;
+        tm.put(TRACE_ID, traceId);
+      }
+    }
+
+    @Override
+    public <C> SpanContext extract(Format<C> format, C carrier) {
+      if (!format.equals(Format.Builtin.HTTP_HEADERS)) {
+        throw new IllegalArgumentException("unsupported format " + format);
+      }
+      String traceId = null;
+      TextMap tm = (TextMap) carrier;
+      Iterator<Entry<String, String>> it = tm.iterator();
+      while (it.hasNext()) {
+        var e = it.next();
+        // 'equalsIgnoreCase' because header name might be lowercase
+        if (e.getKey().equalsIgnoreCase(TRACE_ID)) {
+          traceId = e.getValue();
+          break;
+        }
+      }
+      if (traceId == null) {
+        traceId = newTraceId();
+      }
+      return new SimplePropagatorSpanContext(traceId);
+    }
+
+    @Override
+    public String toString() {
+      return "SimplePropagator";
+    }
+  }
+
+  private static final class SimplePropagatorSpanBuilder implements SpanBuilder {
+
+    private final ScopeManager scopeManager;
+    // storing parent to support the parent being injected via the asChildOf method
+    private SpanContext parent;
+
+    public SimplePropagatorSpanBuilder(ScopeManager scopeManager) {
+      this.scopeManager = scopeManager;
+    }
+
+    @Override
+    public Span start() {
+      if (parent != null) {
+        return new SimplePropagatorSpan(parent);
+      } else if (scopeManager.activeSpan() != null) {
+        return new SimplePropagatorSpan(scopeManager.activeSpan().context());
```

Comment:
This is going to be the common case as more spans show up _within_ Solr.  Could SimplePropagatorSpan and SimplePropagatorSpanContext be combined into one?  Create no new object; the context could *be* the span if all we are doing is propagation.

I even wonder if we need a distinct builder?

Reply:
good suggestion. I merged the 2 but it introduced some minor complexity dealing with the parent context, but not a big issue.